components/causal_ts/src/observer.rs

```diff
@@ -79,14 +158,44 @@ impl<Ts: CausalTsProvider> QueryObserver for CausalObserver<Ts> {
 impl<Ts: CausalTsProvider> RoleObserver for CausalObserver<Ts> {
     /// Observe becoming leader, to flush CausalTsProvider.
     fn on_role_change(&self, ctx: &mut ObserverContext<'_>, role_change: &RoleChange) {
-        if role_change.state == StateRole::Leader {
-            let region_id = ctx.region().get_id();
+        fail_point!("causal_ts_before_on_role_change", |_| ());
+        let region_id = ctx.region().get_id();
+        let is_leader = role_change.state == StateRole::Leader;
+        if is_leader {
```

Comment:
I think this problem can be solved easily by flushing when it becomes candidate. And leader is only needed to be checked when it's singleton.

Reply:
I tried this solution and verified by E2E testing that it works. PTAL https://github.com/tikv/tikv/pull/12573, thanks~